Adding Data (but not Generic) instance

recursion-ninja opened the pull request:
Just adding these instances. Required for consumers of the `text-short` library to derive `Data` & `Generic` instances automatically via `-XDeriveDataTypeable` & `-XDeriveGeneric`, respectively.

hvr commented:
What's the concrete motivation/use-case for having these instances? I usually avoid defining these instances for abstract datatypes as supplying these exposes the internal representation to some degree...

recursion-ninja commented:
Our use case as serialization via the compact regions of the `compact` package. It has a `Typeable` constraint, and nested deep in our object are some `ShortText` values that need `Data` instances to be "glued" together for a `Typeable` instance of the large object to be serialized.

Thanks for dredging up the link from [`Data.Text`](https://github.com/haskell/text/issues/256).

I read through the conversation surrounding the `Data` instance decision of `Text`. I found that [this comment][0] from Neil mirrors my own philosophy on the matter:

> You have to trust people not to break that abstraction, but it's not a simple abstraction to break - it's the moral equivalent of pointer prodding in a std::string, no one breaks it accidentally.

Also, I was a little surprised that [`ByteString`][1], [`ShortByteString`][2], and [`Text`][3] have [`Data`][4] instances, but not `ShortText`. It seemed a bit of an "inconsistent API," though I know that these types are from three separate packages and that there isn't really an API for all the "string type" packages.

[0]: https://mail.haskell.org/pipermail/haskell-cafe/2010-January/072388.html
[1]: http://hackage.haskell.org/package/bytestring-0.10.8.2/docs/Data-ByteString.html#i:ByteString
[2]: http://hackage.haskell.org/package/bytestring-0.10.8.2/docs/Data-ByteString-Short.html#i:ShortByteString
[3]: http://hackage.haskell.org/package/text-1.2.3.1/docs/Data-Text.html#i:Text
[4]: http://hackage.haskell.org/package/base-4.11.0.0/docs/Data-Data.html#t:Data

recursion-ninja commented:
Looks like there's another person hoping to use the `Data` instance here: #7

hvr commented:
@recursion-ninja I'm actually ok with adding a `Data` instance in the style of `Data.Text`; then we'd be consistent with the prior art. Would that satisfy your needs?

recursion-ninja commented:
@hvr yes, that's manner in which I was hoping to use the ` ShortText`  `Data` instance.

recursion-ninja commented:
@hvr any word on merging the PR and uploading a new version of `short-text` to hackage?

hvr commented:
@recursion-ninja well, the PR as-is a) adds a `Generic` instance (which leaks internals), and b) doesn't implement the `Data` instance in the style of `Data.Text` (i.e. by faking a `[Char]` representation)

recursion-ninja commented:
@hvr I misunderstood your previous comment. What you said with regards to the `Data` & `Generic` instances in the pull request makes sense.

Consequently, I removed the `Generic` instance from this pull request and changed the `Data` instance from the standard `deriving` mechanism to the abstracted `[Char]` representation similar to `Text`.

Please take a peek at the updated `Data` instance to make sure it's both sensible and what you had in mind.

sjakobi commented:
Bump @hvr!

I'd also like to have that instance! :)

hvr commented:
Thanks; I'll take it from here
